### ISSUE TITLE
Tableau Evalution

### ISSUE THREAD

**Issue by baiqiushi** [ISG-ICS/cloudberry#688]
# First try by Dengli in 2018

https://docs.google.com/presentation/d/1AIaOQiIZfCOmWXcMv8b-SSFQzGVwg8FahtwL4ASMeiU/edit?usp=sharing

**Comment by JosephZheng1998** [comment]
There are two ways that Tableau connects to data sources and databases that are not in the list of Tableau's default and built-in databases. The first way is through a customized ODBC driver, and the second way is through a Web Data Connector.

An ODBC driver uses the Open Database Connectivity (ODBC) interface by Microsoft that allows applications to access data in database management systems (DBMS) using SQL as a standard for accessing the data. ODBC permits maximum interoperability, which means a single application can access different DBMS. Application end users can then add ODBC database drivers to link the application to their choice of DBMS. Since building an ODBC driver is too complicated, I did not attempt to build an ODBC driver.

The Web Data Connector uses the Tableau Web Data Connector SDK. It essentially allows you to retrieve data from any database or data source of your choice. A WDC is an HTML page with JavaScript code that connects to web data (for example, by means of a RESTful API), converts the data to a JSON format, and passes the data to Tableau. However, it only extracts the data from the data source and does not actually establish a connection to the database. The WDC Tutorial by Tableau provides an HTML and JavaScript template for building a Web Data Connector, and I used these templates to build a Web Data Connector to retrieve data from the MovieDB from CS 122B.

I used a Python script to establish a connection to MySQL, retrieve the data, and convert the data into JSON format. I generated a text file that contains this information. Afterward, I used my JavaScript script to specify all the schemas, tables and table columns, including all the data names and the data types. Then, I load the data from the generated text file and link the JavaScript to the HTML page. Finally, I type in the URL in the HTML page of the WDC to retrieve the data to Tableau for visualization.

In conclusion, we can use Tableau's Web Data Connector to extract data from Cloudberry and put it into Tableau for visualization. However, it is not possible to establish a connection between Tableau and Cloudberry without writing an ODBC driver for Cloudberry.
